server/src/main/java/org/elasticsearch/action/datastreams/autosharding/DataStreamAutoShardingService.java

```diff
@@ -0,0 +1,424 @@
+            );
+
+            return new AutoShardingResult(
+                AutoShardingType.NO_CHANGE_REQUIRED,
+                writeIndex.getNumberOfShards(),
+                writeIndex.getNumberOfShards(),
+                TimeValue.ZERO,
+                writeIndexLoad
+            );
+        }
+
+        return getDecreaseShardsResult(metadata, dataStream, writeIndexLoad, nowSupplier, writeIndex, remainingReduceShardsCooldown);
+    }
+
+    @Nullable
+    private AutoShardingResult getIncreaseShardsResult(
+        DataStream dataStream,
+        double writeIndexLoad,
+        LongSupplier nowSupplier,
+        IndexMetadata writeIndex
+    ) {
+        // increasing the number of shards is calculated solely based on the index load of the write index
+        long optimalShardCount = computeOptimalNumberOfShards(minWriteThreads, maxWriteThreads, writeIndexLoad);
+        if (optimalShardCount > writeIndex.getNumberOfShards()) {
+            TimeValue timeSinceLastAutoShardingEvent = dataStream.getAutoShardingEvent() != null
+                ? dataStream.getAutoShardingEvent().getTimeSinceLastAutoShardingEvent(nowSupplier)
+                : TimeValue.MAX_VALUE;
+
+            TimeValue coolDownRemaining = TimeValue.timeValueMillis(
+                Math.max(0L, increaseShardsCooldown.millis() - timeSinceLastAutoShardingEvent.millis())
+            );
+            logger.debug(
+                "data stream autosharding service recommends increasing the number of shards from [{}] to [{}] after [{}] cooldown for "
+                    + "data stream [{}]",
+                writeIndex.getNumberOfShards(),
+                optimalShardCount,
+                coolDownRemaining,
+                dataStream.getName()
+            );
+            return new AutoShardingResult(
+                AutoShardingType.INCREASE_SHARDS,
+                writeIndex.getNumberOfShards(),
+                Math.toIntExact(optimalShardCount),
+                coolDownRemaining,
+                writeIndexLoad
+            );
+        }
+        return null;
+    }
+
+    /**
+     * Calculates the amount of time remaining before we can consider reducing the number of shards.
+     * This reference for the remaining time math is either the time since the last auto sharding event (if available) or otherwise the
+     * oldest index in the data stream.
+     */
+    private TimeValue getRemainingDecreaseShardsCooldown(Metadata metadata, DataStream dataStream) {
+        Index oldestBackingIndex = dataStream.getIndices().get(0);
+        IndexMetadata oldestIndexMeta = metadata.getIndexSafe(oldestBackingIndex);
+
+        return dataStream.getAutoShardingEvent() == null
+            // without a pre-existing auto sharding event we wait until the oldest index has been created longer than the decrease_shards
+            // cool down period "ago" so we don't immediately after a data stream is created, reduce the number of shards
```

Comment:
Super minor, but missing a word between "we don't" and "immediately" I think, did you mean something like
```suggestion
            // cool down period "ago" so we don't immediately reduce the number of shards after a data stream is created
```
?